### ISSUE TITLE
problem with specific dvbt2 ts

### ISSUE THREAD

**Issue by hagaygo** [kodi-pvr/pvr.hts#431]
Hi

I am using tvheadend 4.3 and kodi 18.x with dvb-t/t2/s/s2 inputs.

All works just fine except one dvb-t2 mux.

The problem happens only on kodi (PC/rpi 2-3 libreelec) and only on "live tv" stream.

When recording from channels from that mux , the playback is just fine.
When viewing live tv stream from that mux (timeshift enabled/disabled) the audio gets out of sync quite quickly and there are noise distortions/drops every 2-5 seconds.

When playing channel from vlc or any other player (using the "pass" profile) the playback is just fine.

So my guess was its related to the kodi addon which uses htsp profile for live tv playback.

I did local stream loopack using ffmpeg and made a video copy but re-encoded the audio to mp3.
This way , the stream starts after about 5-6 seconds , but playback is just fine (and audio is mp3 instead of aac).

So i guess there is something "special" in the transport stream of that mux. 

I uploaded a small ts copy from that mux to :

https://github.com/hagaygo/testfiles/blob/master/DVBT2IL.ts?raw=true

I reproduced the problem by creating an IPTV input with the following code (which keeps the audio as is) :

pipe:///usr/bin/ffmpeg -loglevel error -i https://github.com/hagaygo/testfiles/blob/master/DVBT2IL.ts?raw=true -vcodec copy -acodec copy -metadata service_provider=FFMPEG
  -metadata service_name="test service"  -f mpegts pipe:1

You can try playing that url with vlc and see it plays fine , using kodi tv live stream this channel , the problem should be clear.

Thanks in advance for your work and time.

let me know what else i should supply/check.

**Comment by hagaygo** [comment]
Any chance any one checked/can check this ?

There are other reports on this problem  : 

https://tvheadend.org/issues/5636

https://tvheadend.org/boards/5/topics/33234

**Comment by ksooo** [comment]
Most probably a Kodi problem and not caused by the add-on. The add-on does only pass the stream data from tvh to Kodi, without modifying it. Please open a Kodi issue.

**Comment by hagaygo** [comment]
Thanks for your response,

I would agree if :
1.Playing the same stream on kodi via http url would behave the same (using pass profile)  , but its not , it plays perfectly on kodi.
2.Recording , if i record (on TVH) the stream (using pass , with out any re-encode) and play the recording in kodi , plays perfectly.

Please find the time so you can verify my finding (you wont understand the language but you'll see the audio is out of sync and you will hear audio distortion from time to time)

The issue happens only while playing via htsp/addon on kodi live tv.

**Comment by FernetMenta** [comment]
> Most probably a Kodi problem

I don't think so. vlc does not use tvheadend's demuxer, hence you compare apples with oranges. most likely an issue with the hts protocol in tvh.

**Comment by hagaygo** [comment]
Any update ? 
No one even curious about this behavior with this TS ?
Any thing i can do to help getting this fixed ?

**Comment by hagaygo** [comment]
Tested PR https://github.com/tvheadend/tvheadend/pull/1343 , did not fix this issue.

**Comment by hagaygo** [comment]
Found another problematic TS.

Dump of the ts is on https://github.com/hagaygo/testfiles/blob/master/hevc_ts_htsp_problem.ts?raw=true

Its an HEVC one , which works just fine with "pass" profile , but cant be played via kodi using htsp.

Even the FFMPEG trick does not work with this TS.

Only way to play it on kodi is to start a record and then start playing the record file from "Recording".

Its quite clear there will be "problematic" TS from time to time.

IMHO if we had an option to either not use htsp for viewing channels on the addon (using "pass" profile, and losing timeshift and other functionalities) or better allow us the set "pass" profile by channel (even via xml file for start , no need for GUI) , so we can overcome problematic TS and wont lose functionality on other channels.

**Comment by FernetMenta** [comment]
Unfortunately the hts protocol for streaming is buggy in many aspects and looks like this won't change in the future. It would be great if pvr.hts could support pass profile.
I experience many htsp errors that result in picture corruptions on some channels.

**Comment by CvH** [comment]
@FernetMenta would make the pass profile client side timeshift possible (ofc someone needs to code it)? 
Modern clients have easily 1gb+ RAM available for timeshift available.

**Comment by ksooo** [comment]
> IMHO if we had an option to either not use htsp for viewing channels on the addon (using "pass" profile

Doesn't it help to specify "pass" in the addon settings to achieve what you want?

![screenshot001](https://user-images.githubusercontent.com/3226626/92780181-0b1ead00-f3a3-11ea-82c1-19c43d9467bd.png)

BTW: Did not realize that before, but according to my tvh settings "pass" is even the default. No idea whether I changed that or whether this was always like this. ;-)

<img width="1141" alt="Screenshot 2020-09-10 at 20 23 08" src="https://user-images.githubusercontent.com/3226626/92781582-8c2a7400-f3a4-11ea-9ed5-262d99c77b75.png">

**Comment by hagaygo** [comment]
"pass" is also the default here , but you can check tvh log and see that it uses htsp profile (profile="htsp" in the log when selecting a channel)

when i set pass as profile on the addon setting i get this error :
![image](https://user-images.githubusercontent.com/5015856/92788730-de729180-f3b2-11ea-8298-81d8f65c8473.png)

As far as i know , the addon (currently ?) uses hts protocol only and thus can't really use the "pass" profile.

**Comment by FernetMenta** [comment]
> @FernetMenta would make the pass profile client side timeshift possible (ofc someone needs to code it)?
> Modern clients have easily 1gb+ RAM available for timeshift available.

Yes, that's probably the way to go. TVh's timeshift is as buggy as the htsp streaming protocol with its integrated demuxer. In a first iteration pvr.hts could pass the http streaming url in dynpath of file items (mybe dependent on a setting). Then playback of channels would work like for all other pvr addons using ffmpeg demuxer.

**Comment by FernetMenta** [comment]
fyi https://github.com/kodi-pvr/pvr.hts/pull/471

**Comment by ksooo** [comment]
Is this still an issue with latest Kodi v19 nightly builds? If I get no response within two weeks, I will close this issue.

**Comment by rdorsch** [comment]
I think this issue should be reopened, also #498 seems to suffer from this one. Even if it is not directly related to this addon, but a bug in the tvheadend server or htps protocol, I suggest to reopen it. At least it would have helped me to find it. Since it was closed, I just found it because @hagaygo pointed me to it.

I used

LibreELEC-iMX6.arm-10.0-nightly-20210312-c5d7822-cubox.img.gz

and hit this or a very similar issue. One difference I see is that it affects all stream on kodinerds-iptv for me. But maybe it is easier to hit the issue on imx6?

I also would prefer to have a pass option and timeshift disabled if the pass option is selected. Having it not working at all, I have to use something different than tvheadend :-/ .
